src/org/mozilla/javascript/ScriptableObject.java

```diff
@@ -1597,6 +1613,37 @@ protected void defineOwnProperty(
         }
     }
 
+    /**
+     * Define a property on this object that is implemented using lambda functions. If a property
+     * with the same name already exists, then it will be replaced. This property will appear to the
+     * JavaScript user exactly like any other property -- unlike Function properties and those based
+     * on reflection, the property descriptor will only reflect the value as defined by this
+     * function.
+     *
+     * @param name the name of the function
+     * @param getter a function that returns the value of the property. If null, then the previous
+     *     version of the value will be returned.
+     * @param setter a function that sets the value of the property. If null, then the value will be
+     *     set directly and may not be retrieved by the getter.
+     * @param attributes the attributes to set on the property
+     */
+    public void defineOwnProperty(
+            String name, Supplier<Object> getter, Consumer<Object> setter, int attributes) {
```

Comment:
Do you need to have a `checkValid` flag like the `defineOwnProperty(Context cx, Object id, ScriptableObject desc, boolean checkValid)` method or otherwise determine if you are allowed to replace a property that already exists?

Reply:
I actually renamed this to "defineProperty" to better match the existing "defineProperty" mechanism, which is for Java reflection. That function doesn't do any checking either. I think that since both of these functions are really intended for use in Java code to implement things natively in Java, it's OK to forgo the spec-compliant checking.

Also, the only way we have to check now requires us to create a property descriptor and do a lot of complex comparisions -- a more efficient way to do that would be cool.